### PR TITLE
MIDI-55: VQ-VAE + MIDI

### PR DESCRIPTION
**Pull request by SamuelJanas**
[MIDI-55](https://sacrebleu.atlassian.net/browse/MIDI-55)
[MIDI-59](https://sacrebleu.atlassian.net/browse/MIDI-59)
[MIDI-81](https://sacrebleu.atlassian.net/browse/MIDI-81)
Dataset used in this project at [huggingface](https://huggingface.co/datasets/SneakyInsect/maestro-preprocessed)

[MIDI-55]: https://sacrebleu.atlassian.net/browse/MIDI-55?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MIDI-59]: https://sacrebleu.atlassian.net/browse/MIDI-59?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MIDI-81]: https://sacrebleu.atlassian.net/browse/MIDI-81?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by SamuelJanas** [comment]
Basic training loop implemented. Will be looking at reshaping/changing the architecture to work with 1D data next.

**Comment by SamuelJanas** [comment]
Adapted the code from [VQ-VAE-Speech](https://github.com/swasun/VQ-VAE-Speech).
There are no problems with dimensionality now but the training loss is growing exponentially. 😅 
image
I'm currently working on a fix, most likely the lack of normalization is at fault

**Comment by roszcz** [comment]
Too high learning rate could also cause this 🤔

**Comment by SamuelJanas** [comment]
> Too high learning rate could also cause this 🤔

Hmm, you might be right, I was at 2e-4 for the screenshot above, at 1e-5:
image
It got better already.
Currently training with 1e-6 which worries me a little since it's incredibly small and the results look a little random?
image

**Comment by roszcz** [comment]
Can you showcase a couple of samples of the data? It's good review what exactly is the input and expected output for the model

**Comment by SamuelJanas** [comment]
> Can you showcase a couple of samples of the data? It's good review what exactly is the input and expected output for the model

this was the plan:
image
While debugging the code I found some other problems:
```Python
velocity = x_combined[0, :, 2].detach().cpu().numpy()
velocity_recon = reconstructed_x[0, :, 2].detach().cpu().numpy()
print(f"Velocity: {np.min(velocity)} - {np.max(velocity)}, shape: {velocity.shape}")
print(f"Velocity recon: {np.min(velocity_recon)} - {np.max(velocity_recon)}, shape: {velocity_recon.shape}")
```
The code returns:
```
Velocity: 3.0 - 110.0, shape: (1024,)
Velocity recon: -0.289111465215683 - -0.05111280828714371, shape: (1024,)
```
it seems like either the model fails to find appropriate values or it's a fault of some activation function (I coulndn't find it in the code though)

Since de-normalizing velocity from [-1, 1] should be rather easy I'll check if normalizing data in the loader will make any sense

**Comment by roszcz** [comment]
From what I see you're using a different implementation of VQ-VAE here - I just checked and it looks like the model here has ~25M parameters, while the one you used in the auto-encoder intro project had ~0.6M parameters. 

I understand that you changed implementation in order to fit the use case, butI think it will be better to go back to the previous model, and modify it to work with the MIDI structures we want to model. 

Also, please take a look at discord :)

**Comment by SamuelJanas** [comment]
> From what I see you're using a different implementation of VQ-VAE here - I just checked and it looks like the model here has ~25M parameters, while the one you used in the auto-encoder intro project had ~0.6M parameters.
> 
> I understand that you changed implementation in order to fit the use case, butI think it will be better to go back to the previous model, and modify it to work with the MIDI structures we want to model.

I changed some of the parameters in the config to better fit the size of the data, the model now has ~7M parameters and can be easily downsized further with additional tweaks in the config. The training looks much better afterwards, probably as you noticed it was simply too big. Additionaly I normalized the velocity, which should be a loss-less conversion and changed the start times to treat each track like it's starting from the beginning (subtracted `start[0]` from all values). I believe we can achieve better results with a proper config.

**Comment by SamuelJanas** [comment]
To answer @roszcz comment from [MIDI-58](https://sacrebleu.atlassian.net/browse/MIDI-58), the [dataset](https://huggingface.co/datasets/SneakyInsect/maestro-preprocessed) is built from [maestro-v1-sustain](https://huggingface.co/datasets/roszcz/maestro-v1-sustain)
In this PR we are mainly focusing on `start`, `duration` and `velocity`, here's the change list compared to maestro:

- duration: Unchanged
- pitch: Unchanged
- velocity: Normalized to [-1, 1]
- start: Adjusted to treat each rolling window as a separate track, starting at 0.0 each row
- end: Adjusted accordingly
- We use a rolling window of size 60 with a hop size of 20 for increased coverage.

Relevant commits: [d095ef4](https://github.com/Nospoko/vqvae-midi/pull/1/commits/d095ef4b81649239cd06640243427312c041730b), [93509bb](https://github.com/Nospoko/vqvae-midi/pull/1/commits/93509bb0a5a210d09a30d9e2d74cc5e5f4e7d4f4), [192252e](https://github.com/Nospoko/vqvae-midi/pull/1/commits/192252e182de2fab0a8b7b7c6450997e9da93125)

[MIDI-58]: https://sacrebleu.atlassian.net/browse/MIDI-58?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by SamuelJanas** [comment]
As a quick update, this is how the single batch overfit works for now:
Overfitting a single batch
It has all the problem I mentioned on the call, there is a clearly visible problem with VQ-Loss (console logs when running locally), I'll be reading some materials on VQ-VAE to hopefully come up with the fix.



Will also be changing start to dstart, this might switch the VAE focus towards velocity and duration from the start value.
Another idea would be to just encode the velocity to later on combine it with @JasiekKaczmarczyk's work

**Comment by roszcz** [comment]
To me this looks pretty far from overfitting - if this is what you're getting from a multi-million parameter model on a single batch, it's possible that there are some more fundamental bugs preventing the model from learning at all 🤔

**Comment by SamuelJanas** [comment]
> To me this looks pretty far from overfitting - if this is what you're getting from a multi-million parameter model on a single batch, it's possible that there are some more fundamental bugs preventing the model from learning at all 🤔

The values on the screenshot are just samples from the beginning of training, reconstruction loss does go down to approx. 0.3 eventually, but you're right the convergance should be faster on one batch. There might be some funamental bugs somewhere, still under investigation 🔬

**Comment by SamuelJanas** [comment]
I've created better visualizations to look at the problem. All experiments done on default `config_single_batch.yaml`, with changes to learning rate.
|  lr  | losses | values |
|:-----:|------|--------|
| $1e-4$ |losses0 0001|values_0 0001|
| $1e-5$ |losses1e-05|values_1e-05|
| $1e-6$ |losses1e-06|values_1e-06|

Trained the low lr version for 3x longer and expect it to spike similarily to the other ones.
Interesting to see how it finds the average for dstart, zeros velocity(?) since it's normalized [-1, 1] the values are somewhere around the 0 and works mainly with duration. I'm trying to optimize the parameters first to see how low we can get with the reconstruction loss and I'll be looking at the VQ next

A model with cfg changed the following way:
```
  num_hiddens: 512
  num_residual_layers: 4
  num_residual_hiddens: 256
  embedding_dim: 90
  num_embeddings: 32
```
Can achieve good-ish results, but the embedding is compressed only twice which could probably be improved:
values_0 0001

**Comment by roszcz** [comment]
When you have a very large model and a very small batch, you can expect the model to memorise the target data verbatim, which makes the loss go to zero. Looking at those charts it looks to me like a model that didn't learn anything, other than the mean values for the output distribution 🤔

**Comment by SamuelJanas** [comment]
> When you have a very large model and a very small batch, you can expect the model to memorise the target data verbatim, which makes the loss go to zero. Looking at those charts it looks to me like a model that didn't learn anything, other than the mean values for the output distribution 🤔

I agree that seems a little sketchy, but I believe the model is capable of memorizing verbatim for one batch. It is simply very frial to hyperparameters, I made some changes:
1. Normalization of velocity is now on [0-1] range, just made more sense to me to have all values positive.
2. I've learned that changing the amount of embeddings to a lower one worked better for training single batch. My intuition is that with higher embeddings number it needs more time to set on some values
3. I've managed to train the model below with `embedding_dim = 45`, which is half of the one earlier and still got better results

values_0 0002
A thing worth noting is that the learning rate can be much higher with properly chosen parameters,

losses0 0002

Training for those charts was done over 2000 epochs or ~1 minute for one batch
@roszcz give those results a look and let me know if they look promising to you. If you have any suggestions I would greatly appreciate them. Also let me know If I should focus on tweaking the parameters, which seems to have worked for now or maybe I should look deeper into the architecture to possibly make some changes so it's more robust to them

**Comment by JasiekKaczmarczyk** [comment]
@SamuelJanas are you using normalization (BatchNorm/GroupNorm) between ConvBlocks? If not, that's probably the reason why training is so unstable and so dependent on hyperparameters.

If you have some time I really recommend watching this: https://www.youtube.com/watch?v=P6sfmUTpUmc, especially section about BatchNorm

**Comment by SamuelJanas** [comment]
> @SamuelJanas are you using normalization (BatchNorm/GroupNorm) between ConvBlocks? If not, that's probably the reason why training is so unstable and so dependent on hyperparameters.

That helped a lot, thanks! I've started using BatchNorm in-between the conv layers in the encoder part of the model, using it anywhere else(ResBlocks, decoder) ended up worsening the training. But overall, the model is more robust to hyperparameters so big improvements, thanks again.

> If you have some time I really recommend watching this: https://www.youtube.com/watch?v=P6sfmUTpUmc, especially section about BatchNorm

After watching it (great vid btw) I've managed to fix a small inaccuracy, kaiming initialization was set up for leaky_relu when model is using relu. In addition to that I see that some improvements can be done on init of training.

I trained another model with more trainable parameters. My reasoning was that if it cannot overfit for a single batch, maybe the model isn't complex enough. The results are good imo.
values_0 0001
Most of the time, it's on point with original data, some jumps here and there, struggling with duration the most, not sure why yet. Overall I believe it's rather good(?)
losses0 0001
Visible that the training is more stable
I'm going to look at all the other tracks to see some patterns.

**Comment by SamuelJanas** [comment]
It seems like the track I was benchmarking on is an especially difficult one. Here are some examples that show that the model isn't that bad. 😅 
16_values_0 0001

20_values_0 0001

26_values_0 0001

**Comment by roszcz** [comment]
Nice! 🚀 Can you start including listening of those sequences to your testing workflow? I don't mean anything intertwined with the training code, just tools and habit to generate a pianoroll and mp3s of those sequences. 

Could be interesting to see if the "difficult" one actually sounds worse with those reconstruction errors :)

File: error_handling/color_print.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import sys
+
+
+class ColorPrint(object):
```
Comment (roszcz) [inline error_handling/color_print.py:4]: Check out rich library: https://github.com/Textualize/rich if you like printing colors :)

File: model/vector_quantizer.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                "e_latent_loss": e_latent_loss.item(),
+                "q_latent_loss": q_latent_loss.item(),
+                "commitment_loss": commitment_loss.item(),
+                "vq_loss": vq_loss.item(),
+            },
+            encoding_distances,
+            embedding_distances,
+            frames_vs_embedding_distances,
+            concatenated_quantized,
+        )
```
Comment (roszcz) [inline model/vector_quantizer.py]: Can we use a dict or maybe a dataclass for this? 20 element tuples are untraceable

**Review by roszcz — COMMENTED** [review]
Please use type hints and named parameters as much as possible - I'm not so young anymore and struggle to read code without it 🤓

**Comment by JasiekKaczmarczyk** [comment]
> > @SamuelJanas are you using normalization (BatchNorm/GroupNorm) between ConvBlocks? If not, that's probably the reason why training is so unstable and so dependent on hyperparameters.
> 
> That helped a lot, thanks! I've started using BatchNorm in-between the conv layers in the encoder part of the model, using it anywhere else(ResBlocks, decoder) ended up worsening the training. But overall, the model is more robust to hyperparameters so big improvements, thanks again.
> 
> > If you have some time I really recommend watching this: https://www.youtube.com/watch?v=P6sfmUTpUmc, especially section about BatchNorm
> 
> After watching it (great vid btw) I've managed to fix a small inaccuracy, kaiming initialization was set up for leaky_relu when model is using relu. In addition to that I see that some improvements can be done on init of training.
> 
> I trained another model with more trainable parameters. My reasoning was that if it cannot overfit for a single batch, maybe the model isn't complex enough. The results are good imo. values_0 0001 Most of the time, it's on point with original data, some jumps here and there, struggling with duration the most, not sure why yet. Overall I believe it's rather good(?) losses0 0001 Visible that the training is more stable I'm going to look at all the other tracks to see some patterns.

It should also work better when using normalization in decoder and resnet blocks but the last layer of decoder (just like you have in encoder) should only have conv without normalization and activation afterwards, that's probably why it was worsening your training.

So encoder/decoder would look something like this:
Conv
Norm
Activation
Conv
Norm
Activation
...
Conv

For example:
```python
class ConvBlock(nn.Module):
    def __init__(self, in_channels: int, out_channels: int, upscale: bool = False, **kwargs):
        super().__init__()

        self.upscale = upscale

        self.conv = nn.Sequential(
            nn.Conv2d(in_channels, out_channels, **kwargs),
            nn.BatchNorm2d(out_channels),
            nn.ReLU()
        )

    def forward(self, x: torch.Tensor):
        if self.upscale:
            x = F.interpolate(x, scale_factor=2)
        
        return self.conv(x)

class Encoder(nn.Module):
    def __init__(self, in_channels: int, out_channels: int, <other params>):
        super().__init__()

        self.encoder = nn.Sequential(
            ConvBlock(in_channels, channels[0], upscale=False/True, <other params>) # N times
            # ...
            Conv1d(channels[-1], out_channels, <other params>)
        )

    def forward(self, x: torch.Tensor):
        return self.encoder(x)
```

**Comment by SamuelJanas** [comment]
> It should also work better when using normalization in decoder and resnet blocks but the last layer of decoder (just like you have in encoder) should only have conv without normalization and activation afterwards, that's probably why it was worsening your training.
> 
> So encoder/decoder would look something like this: Conv Norm Activation Conv Norm Activation ... Conv

I did just that but that didn't really help. I'll give it some more tries, maybe I used something in a wrong way. I'll let you know if it helps though!

File: error_handling/color_print.py (hunk 1 above)
Comment (SamuelJanas) [reply error_handling/color_print.py:4]: That looks *shiny*. 😄 The ColorPrint class was copied from the repo I mentioned, but I'll use rich when dealing with colors from now on

File: model/vector_quantizer.py (hunk 2 above)
Comment (SamuelJanas) [reply model/vector_quantizer.py]: changed to dict

**Comment by SamuelJanas** [comment]
In a teamwork fashion I have stolen @JasiekKaczmarczyk  [code](https://github.com/Nospoko/midi-diffusion/blob/MIDI-60/midi-diffusion/evals/compare_original_and_generated.py#L32) for MIDI generation. 

> Could be interesting to see if the "difficult" one actually sounds worse with those reconstruction errors :)

Surprisingly, the reconstruction is very good for the overfited batch, here are some samples:
["difficult one"](https://audio.jukehost.co.uk/P1MMoxasAmQtEwK9iZo19GprEROgUmph) and its' [reconstruction](https://audio.jukehost.co.uk/bSz05ynPayeELQtVy7eZ7GI5qcwqDTRB)

[original 2](https://audio.jukehost.co.uk/MRox8t8u5zbEi5IGSlmb6Eh8ldcmLUdA) - [reconstructed 2](https://audio.jukehost.co.uk/KgOuOqAdYSW1Oq7ncLwKwv1VPhK3ZaXE)

I can link some other ones for you but they all sound more or less the same as the original
What I find interesting is that, the reconstruction is better for more harmonious pieces indicating that they are more 'structured'? Following that the difficulty of the benchmarking piece makes sense after hearing it, the notes are short and sharp

**Comment by roszcz** [comment]
I wouldn't pay too much attention to results on a single/overfitted batch, it's mostly useful to identify major problems with the model and training code, not to review the quality of the processing.

I think it's ok to focus on the full dataset and training now :)

**Comment by SamuelJanas** [comment]
Just a quick update, when fixing the VQ, I found [this issue](https://github.com/mkotha/WaveRNN/issues/1), according to which VQ with EMA trains better. Before moving onto the entire dataset I've tried it on a single batch and that would be correct. In the same amount of training, even on smaller model `num_residual_hiddens = 128` down from `512`,  Recon_loss went down below `0.001`.

**Comment by SamuelJanas** [comment]
## overview 
[Run info](https://wandb.ai/sjanas/MIDI%20VQ-VAE/runs/l7xvpf7a?workspace=user-sneakyinsect)
Trained model on the entire dataset, left it training for ~1hour or 17 epochs. Used VQ version with EMA, proved to be better as mentioned in previous comment. Here are the most important parameters for the run:
```yaml
  # Encoder-Decoder details
  num_hiddens: 512
  num_residual_layers: 2
  num_residual_hiddens: 512
  use_kaiming_normal: True
  # VQ details
  embedding_dim: 60 # We might want to lower this value in the future
  num_embeddings: 512 # K, needs empirical tuning
  commitment_cost: 0.25
  decay: 0.99
  ```
The results are promising, I believe that with more training time it could work better similarly to how some drops in loss occured in overfitting for a batch.

## results 
2 randomly chosen pieces from validation set:
- [original](https://audio.jukehost.co.uk/QDdBmElWdiotH6ORT5La8gLR9ePb9Q4I) - [reconstructed](https://audio.jukehost.co.uk/oyITVonrKnjS0HDjdRyTDmRbHqVycXUU)
14_values_1e-06

- [original](https://audio.jukehost.co.uk/bwEs9u8DX6fN3KVA9HaMFOfSD6Lei5cB) - [reconstructed](https://audio.jukehost.co.uk/hE3joFFSvMgDMliifihTmMUwAN23hyGB)
1_values_1e-06

It's far from perfect but does the job decently, I have to look at the K value in the config and possibly make the model larger? Let me know your thoughts. 😃

**Comment by SamuelJanas** [comment]
I've created a dashboard  for you to explore the model. 
image
It's a WIP, it should probably be moved to column view similarly to @WojciechMat dashboard. Works not too fast but managable but pre-generating all the data would weight too much I believe. 😅 

You can either train your own model with `python train.py` or use my pre-trained one (165 epochs, ~12 hours) from [here](https://huggingface.co/SneakyInsect/VQ-VAE-MIDI/blob/main/2023_09_06_21_32_all_data165.pt).
In case you train your own model make sure to put the correct checkpoint in this line.
```Python
    ckpt_path = "checkpoints/2023_09_06_21_32_all_data165.pt"
```

**Comment by SamuelJanas** [comment]
I should probably mention that the K value is probably too small. Some reconstructed tracks sound like they are from a different composer but categorised by VQ into the wrong cluster. It's something I'll look more into

**Comment by SamuelJanas** [comment]
Added value information about the loss so that we know, whether the results is a good one or not. This is how the dashboard looks after some small design changes:
image

I have also uploaded a trained checkpoint for higher K value, it had less training time but achieved similar results. Training was less stable, that sort of makes sense, there is more movement in the VQ part, thus the loss can jump from time to time.
model, K=1024: https://huggingface.co/SneakyInsect/VQ-VAE-MIDI/blob/main/2023_09_08_00_42_all_data127.pt
run information: https://wandb.ai/sjanas/MIDI%20VQ-VAE/runs/yhyvws98?workspace=user-sneakyinsect

Generally speaking the model sounds like someone who knows the notes but that's just about it. Some recreations aren't bad some sound awful. Let me know if you have any other suggestions to dashboard/experiment.

**Comment by roszcz** [comment]
@SamuelJanas, for the evals, can you look into using huggingface as a storage for checkpoints? You can see how this could work in our diffusion repo: https://github.com/Nospoko/midi-diffusion/blob/0b3b3743c0900d53adb7af374e6f2624f51cf93a/evals/compare_original_and_generated.py#L173 :)

By default, evals could download a checkpoint pre-trained by you, and if I have my own local versions, it should be possible to use those instead

**Comment by SamuelJanas** [comment]
> @SamuelJanas, for the evals, can you look into using huggingface as a storage for checkpoints? You can see how this could work in our diffusion repo: https://github.com/Nospoko/midi-diffusion/blob/0b3b3743c0900d53adb7af374e6f2624f51cf93a/evals/compare_original_and_generated.py#L173 :)
> 
> By default, evals could download a checkpoint pre-trained by you, and if I have my own local versions, it should be possible to use those instead

Done 😄

File: train.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        wandb.log(
+            {
+                "test/Average Test Loss": avg_test_loss,
+                "test/Reconstruction Loss": avg_reconstruction_loss,
+            }
+        )
+        print(f"Epoch {epoch}, Average Test Loss: {avg_test_loss}")
+
+
+def train(model: VQVAE, train_loader: DataLoader, optimizer, criterion, cfg: DictConfig, epoch: int):
```
Comment (roszcz) [inline train.py]:
```suggestion
def train_step(model: VQVAE, train_loader: DataLoader, optimizer, criterion, cfg: DictConfig, epoch: int):
```

`train` shows up a lot as a string, and grepping around becomes difficult :)

File: train.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+def set_seed(seed):
+    # https://pytorch.org/docs/stable/notes/randomness.html
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
+    np.random.seed(seed)
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
+
+
+def test(model: VQVAE, test_loader: DataLoader, criterion, cfg: DictConfig, epoch: int):
```
Comment (roszcz) [inline train.py]:
```suggestion
def test_step(model: VQVAE, test_loader: DataLoader, criterion, cfg: DictConfig, epoch: int):
```

Same as with `train_step`

File: configs/config.yaml [hunk 5]
```diff
@@ -0,0 +1,42 @@
+train:
+  batch_size: 64
+  epochs: 100
+  log_interval: 10
+  lr: 1e-6
+
+logger:
+  results_path: "results/"
+  checkpoint_path: "checkpoints/"
```
Comment (roszcz) [inline configs/config.yaml:9]: We should take care to ensure existence of this folder - you can either track some .keep file with git, or create the folder from python level before training starts

File: configs/config.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  enable_wandb: False
```
Comment (roszcz) [inline configs/config.yaml]:
```suggestion
  enable_wandb: True
```

I think it's better to log everything, unless you have a specific reason not to :)

**Review by roszcz — APPROVED** [review]

File: configs/config.yaml (hunk 6 above)
Comment (SamuelJanas) [reply configs/config.yaml]: Copied from config in the previous repo, wasn't using it. Deleted :)

**Pull request merged**